avl-tree/src/lib.rs

```diff
@@ -0,0 +1,589 @@
+        }
+
+        self.delete(node_index);
+        self.rebalance(path);
+
+        Some(value)
+    }
+
+    fn balance_factor(&self, left: u32, right: u32) -> i32 {
+        // safe to convert to i32 since height will be at most log2(capacity)
+        let left_height = if left != SENTINEL {
+            self.get_field(left, Field::Height) as i32 + 1
+        } else {
+            0
+        };
+        let right_height = if right != SENTINEL {
+            self.get_field(right, Field::Height) as i32 + 1
+        } else {
+            0
+        };
+
+        left_height - right_height
+    }
+
+    fn left_rotate(&mut self, index: u32) -> u32 {
+        let right = self.get_field(index, Field::Right);
+        let right_left = self.get_field(right, Field::Left);
+
+        self.set_field(index, Field::Right, right_left);
+        self.set_field(right, Field::Left, index);
+
+        right
+    }
+
+    fn right_rotate(&mut self, index: u32) -> u32 {
+        let left = self.get_field(index, Field::Left);
+        let left_right = self.get_field(left, Field::Right);
+
+        self.set_field(index, Field::Left, left_right);
+        self.set_field(left, Field::Right, index);
+
+        left
+    }
+
+    fn update_height(&mut self, index: u32) {
+        let left = self.get_field(index, Field::Left);
+        let right = self.get_field(index, Field::Right);
+
+        let height = if (left + right) == 0 {
+            0
+        } else {
+            let left_height = if left != SENTINEL {
+                self.get_field(left, Field::Height)
+            } else {
+                0
+            };
+            let right_height = if right != SENTINEL {
+                self.get_field(right, Field::Height)
+            } else {
+                0
+            };
+
+            max(left_height, right_height) + 1
+        };
+
+        self.set_field(index, Field::Height, height);
+    }
+
+    fn delete(&mut self, node: u32) {
+        self.allocator.clear_register(node, Field::Left as u32);
+        self.allocator.clear_register(node, Field::Right as u32);
+        self.allocator.clear_register(node, Field::Height as u32);
+        self.allocator.remove_node(node);
+    }
+
+    fn rebalance(&mut self, path: Vec<Ancestor>) {
+        for (parent, branch, child) in path.iter().rev() {
+            let left = self.get_field(*child, Field::Left);
+            let right = self.get_field(*child, Field::Right);
+
+            let balance_factor = self.balance_factor(left, right);
+
+            let index = if balance_factor > 1 {
+                let left_left = self.get_field(left, Field::Left);
+                let left_right = self.get_field(left, Field::Right);
+                let left_balance_factor = self.balance_factor(left_left, left_right);
+
+                if left_balance_factor < 0 {
+                    let index = self.left_rotate(left);
+                    self.set_field(*child, Field::Left, index);
```

Comment:
I see, this is where you correct the pointers from the above question

Reply:
Indeed, the parent is updated outside the `*-rotate` functions.